File: flutter_app/lib/utils/backup.dart
import 'dart:convert';

/// Generate CSV content from semester data.
///
/// The output includes one header row and one data row per course, grouped by
/// semester. A summary row with the overall CGPA is appended at the end.
String generateCSV(
  List<SemesterCSVData> semesters,
  double cgpa,
) {
  final rows = <String>[];
  rows.add('Semester,Course,Credits,Grade,Grade Point');

  for (final sem in semesters) {
    for (final course in sem.courses) {
      rows.add([
        _csvEscape(sem.name),
        _csvEscape(course.name),
        course.credits.toString(),
        _csvEscape(course.grade),
        course.gradePoint.toString(),
      ].join(','));
    }
  }

  rows.add('');
  rows.add('Overall CGPA,$cgpa');

  return rows.join('\n');
}

/// Escape a CSV field value by wrapping in quotes when it contains special characters.
String _csvEscape(String value) {
  if (value.contains(',') || value.contains('"') || value.contains('\n')) {
    return '"${value.replaceAll('"', '""')}"';
  }
  return value;
}

/// Encode app data as a JSON string for backup.
String exportBackupJson(Map<String, dynamic> data) {
  return const JsonEncoder.withIndent('  ').convert(data);
}

/// Parse a JSON backup string. Returns null if the content is not valid JSON.
Map<String, dynamic>? parseBackupJson(String content) {
  if (content.trim().isEmpty) return null;
  try {
    final parsed = jsonDecode(content);
    if (parsed is Map<String, dynamic>) {
      return parsed;
    }
    return null;
  } catch (_) {
    return null;
  }
}

/// Data model for CSV export
class SemesterCSVData {
  final String name;
  final List<CourseCSVData> courses;

  const SemesterCSVData({required this.name, required this.courses});
}

class CourseCSVData {
  final String name;
  final int credits;
  final String grade;
  final double gradePoint;

  const CourseCSVData({
    required this.name,
    required this.credits,
    required this.grade,
    required this.gradePoint,
  });
}

File: client/src/engine/backup.ts
/**
 * Backup, restore, and CSV export utilities.
 *
 * `exportBackup` and `parseBackupFile` depend on browser APIs (Blob, File)
 * and are intended for use in the client only.
 */

/**
 * Export all app data as a downloadable JSON file.
 */
export function exportBackup(data: object): void {
  const json = JSON.stringify(data, null, 2);
  const blob = new Blob([json], { type: "application/json" });
  const url = URL.createObjectURL(blob);
  const a = document.createElement("a");
  a.href = url;
  a.download = `cgpa-backup-${new Date().toISOString().slice(0, 10)}.json`;
  document.body.appendChild(a);
  a.click();
  document.body.removeChild(a);
  URL.revokeObjectURL(url);
}

/**
 * Import a backup from a JSON file.
 * Returns the parsed data or null if the file is empty / not valid JSON.
 */
export function parseBackupFile(file: File): Promise<object | null> {
  return new Promise((resolve) => {
    const reader = new FileReader();
    reader.onload = () => {
      try {
        const text = reader.result as string;
        if (!text || text.trim().length === 0) {
          resolve(null);
          return;
        }
        const parsed = JSON.parse(text);
        if (typeof parsed === "object" && parsed !== null) {
          resolve(parsed);
        } else {
          resolve(null);
        }
      } catch {
        resolve(null);
      }
    };
    reader.onerror = () => resolve(null);
    reader.readAsText(file);
  });
}

/**
 * Generate CSV content from semester data.
 *
 * The output includes one header row and one data row per course, grouped by
 * semester. A summary row with the overall CGPA is appended at the end.
 */
export function generateCSV(
  semesters: {
    name: string;
    courses: {
      name: string;
      credits: number;
      grade: string;
      gradePoint: number;
    }[];
  }[],
  cgpa: number
): string {
  const rows: string[] = [];
  rows.push("Semester,Course,Credits,Grade,Grade Point");

  for (const sem of semesters) {
    for (const course of sem.courses) {
      rows.push(
        [
          csvEscape(sem.name),
          csvEscape(course.name),
          course.credits,
          csvEscape(course.grade),
          course.gradePoint,
        ].join(",")
      );
    }
  }

  rows.push("");
  rows.push(`Overall CGPA,${cgpa}`);

  return rows.join("\n");
}

/** Escape a CSV field value by wrapping in quotes when it contains special characters. */
function csvEscape(value: string): string {
  if (value.includes(",") || value.includes('"') || value.includes("\n")) {
    return `"${value.replace(/"/g, '""')}"`;
  }
  return value;
}
